### PR TITLE
[Security] Rework firewall's access denied rule

### PR DESCRIPTION
**Pull request by dimabory**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no 
| BC breaks?    | no     
| Deprecations? | no 
| Tests pass?   | yes    
| Fixed tickets | ~~#30099~~, #28229
| License       | MIT
| Doc PR        | 

Follow tickets provided above to reproduce bugs. (there are also some project examples)

~~In addition, I'm looking for someone who knows an answer to [this](https://github.com/symfony/symfony/issues/30099#issuecomment-468693492) regarding rework in this PR.~~

### CODE REVIEW

File: src/Symfony/Component/Security/Http/Tests/Firewall/ExceptionListenerTest.php [hunk 1]
```diff
@@ -193,8 +273,14 @@ private function createEvent(\Exception $exception, $kernel = null)
         return new GetResponseForExceptionEvent($kernel, Request::create('/'), HttpKernelInterface::MASTER_REQUEST, $exception);
     }
 
-    private function createExceptionListener(TokenStorageInterface $tokenStorage = null, AuthenticationTrustResolverInterface $trustResolver = null, HttpUtils $httpUtils = null, AuthenticationEntryPointInterface $authenticationEntryPoint = null, $errorPage = null, AccessDeniedHandlerInterface $accessDeniedHandler = null)
-    {
+    private function createExceptionListener(
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/Tests/Firewall/ExceptionListenerTest.php]: must be kept on one line

File: src/Symfony/Component/Security/Http/Tests/Firewall/ExceptionListenerTest.php [hunk 2]
```diff
@@ -147,16 +152,69 @@ public function testAccessDeniedExceptionNotFullFledged(\Exception $exception, \
     {
         $event = $this->createEvent($exception);
 
-        $tokenStorage = $this->getMockBuilder('Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface')->getMock();
-        $tokenStorage->expects($this->once())->method('getToken')->will($this->returnValue($this->getMockBuilder('Symfony\Component\Security\Core\Authentication\Token\TokenInterface')->getMock()));
-
-        $listener = $this->createExceptionListener($tokenStorage, $this->createTrustResolver(false), null, $this->createEntryPoint());
+        $listener = $this->createExceptionListener(
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/Tests/Firewall/ExceptionListenerTest.php]: please revert the change and similar ones in the PR

Comment (dimabory) [reply src/Symfony/Component/Security/Http/Tests/Firewall/ExceptionListenerTest.php]: moved everything on one line

Comment (dimabory) [reply src/Symfony/Component/Security/Http/Tests/Firewall/ExceptionListenerTest.php]: But I'm still wondering why `php-cs-fixer` doesn't align it automatically

**Comment by curry684** [comment]
Please use a single PR per issue next time, easier to review.

Fixes themself are fine.

Status: reviewed

**Comment by curry684** [comment]
Actually it shouldn't be based on `master` as it's a bugfix. I'll see if I can rebase it on 3.4.

**Comment by curry684** [comment]
@chalasr I just discussed with @michaelcullum that I think we should consider merging it into `master` anyway despite being a bugfix, as older applications may actually be accidentally relying on the broken behavior.

So I'd prefer putting it into `master` instead and documenting it as a B/C breaking bugfix in the upgrade guide.

**Comment by nicolas-grekas** [comment]
Let's add some notes in the CHANGELOG.md file of the component, and give some instructions in UPGRADE-4.3.md then

**Comment by curry684** [comment]
@dimabory can you pick that up?

**Comment by dimabory** [comment]
@curry684, @nicolas-grekas  I'm not sure about B/C breaking. Can you guide me on what exactly can be broken with this change?

**Comment by curry684** [comment]
The change may cause the AccessDeniedHandler to be executed now in cases where it wasn't before, and applications may depend on the previous erroneous behavior to some degree.

Technically everything that changes existing behavior is to be considered BC breakage, but in this case the old behavior was technically wrong and out of spec, hence why we're accepting the fix in a minor release, but not backporting it. But the PR should mention it in the upgrade guide.

**Comment by chalasr** [comment]
Looking at the patch and the fixed ticket, I think we should merge in 3.4, every bugfix breaks someone's behavior.
Calling the access denied handler doesn't sound harmful, is it?

**Comment by curry684** [comment]
I'm mainly hesitant because it'll be a patch level change in low level security behavior. I'm hard pressed to find a harmful scenario but I prefer to err on the side of caution where access control is involved...  😉

**Comment by dimabory** [comment]
So what's the base branch `3.4` or `master`?

**Comment by curry684** [comment]
3.4.

Technically it's correct for a bugfix, I won't object to it and @chalasr is the boss of security component  😉

**Review by chalasr — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @dimabory.

**Comment by dimabory** [comment]
Excuse me, guys, but I didn't update `CHANGELOG.md` and `UPGRADE.md` (like mentioned @curry684).

**Comment by fabpot** [comment]
@dimabory As this is a bug fix, there is no need to update the CHANGELOG/UPGRADE files.

**Comment by JarJak** [comment]
@fabpot There probably is a need, see https://github.com/symfony/symfony/issues/31136

File: src/Symfony/Component/Security/Http/Firewall/ExceptionListener.php [hunk 3]
```diff
@@ -131,8 +131,6 @@ private function handleAccessDeniedException(GetResponseForExceptionEvent $event
             } catch (\Exception $e) {
                 $event->setException($e);
             }
-
-            return;
```
Comment (stof) [inline src/Symfony/Component/Security/Http/Firewall/ExceptionListener.php]: this now calls the accessDeniedHandler even when we call the entry point, which looks weird.

Comment (curry684) [reply src/Symfony/Component/Security/Http/Firewall/ExceptionListener.php]: Good point, but it's pretty much technically in line given the rest of the implementation (it also logs at 123, same thing). Not really trivial to fix given how the flow works...

**Comment by andrew-demb** [comment]
These changes make log with message "Access denied, the user is neither anonymous, nor remember-me." lying, because token may be anonymous on `!$this->authenticationTrustResolver->isFullFledged($token)` (https://github.com/chalasr/symfony/blob/fd1408b13869a381fbebf9b9967f7a80e8b141d3/src/Symfony/Component/Security/Http/Firewall/ExceptionListener.php#L121).

**Comment by andrew-demb** [comment]
Also with this PR symfony call access denied handler without any context, that response already generated for anonymous token.

I think it's BC break.

**Comment by JarJak** [comment]
@andrew-demb fix is coming, see https://github.com/symfony/symfony/pull/31142
